validator.go

```diff
@@ -0,0 +1,267 @@
+package jwt
+
+import (
+	"crypto/subtle"
+	"time"
+)
+
+// Validator is the core of the new Validation API. It can either be used to
+// modify the validation used during parsing with the [WithValidator] parser
+// option or used standalone to validate an already parsed [Claim]. It can be
+// further customized with a range of specified [ValidatorOption]s.
+type Validator struct {
+	// leeway is an optional leeway that can be provided to account for clock skew.
+	leeway time.Duration
+
+	// timeFunc is used to supply the current time that is needed for
+	// validation. If unspecified, this defaults to time.Now.
+	timeFunc func() time.Time
+
+	// verifyIat specifies whether the iat (Issued At) claim will be verified.
+	// According to https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6 this
+	// only specifies the age of the token, but no validation check is
+	// necessary. However, if wanted, it can be checked if the iat is
+	// unrealistic, i.e., in the future.
+	verifyIat bool
+
+	// expectedAud contains the audience this token expects. Supplying an empty
+	// string will disable aud checking.
+	expectedAud string
+
+	// expectedIss contains the issuer this token expects. Supplying an empty
+	// string will disable iss checking.
+	expectedIss string
+
+	// expectedSub contains the subject this token expects. Supplying an empty
+	// string will disable sub checking.
+	expectedSub string
+}
+
+// CustomClaims represents a custom claims interface, which can be built upon the integrated
+// claim types, such as map claims or registered claims.
+type CustomClaims interface {
+	// CustomValidation can be implemented by a user-specific claim to support
+	// additional validation steps in addition to the regular validation.
+	CustomValidation() error
+}
+
+func NewValidator(opts ...ValidatorOption) *Validator {
+	v := &Validator{}
+
+	// Apply the validator options
+	for _, o := range opts {
+		o(v)
+	}
+
+	return v
+}
+
+// Validate validates the given claims. It will also perform any custom validation if claims implements the CustomValidator interface.
+func (v *Validator) Validate(claims Claims) error {
+	var now time.Time
+	vErr := new(ValidationError)
+
+	// Check, if we have a time func
+	if v.timeFunc != nil {
+		now = v.timeFunc()
+	} else {
+		now = time.Now()
+	}
+
+	// We always need to check the expiration time, but the claim itself is OPTIONAL
+	if !v.VerifyExpiresAt(claims, now, false) {
+		vErr.Inner = ErrTokenExpired
+		vErr.Errors |= ValidationErrorExpired
```

Comment:
What do you think about these operators, can we remove them in favor of more verbose but simplified logic?

Reply:
Oh yes for sure, outside the scope of this PR. But one of the things I'd love as part of the `v5` is cleaning up these errors.

I wasn't sure if there was another place we could keep a bucket/wishlist of `v5` items?

Reply:
Here: https://github.com/golang-jwt/jwt/pull/234